mlir/lib/Dialect/MQTOpt/Transforms/ElidePermutations.cpp

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2023 - 2025 Chair for Design Automation, TUM
+ * Copyright (c) 2025 Munich Quantum Software Company GmbH
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * Licensed under the MIT License
+ */
+
+#include "mlir/Dialect/Common/Compat.h"
+#include "mlir/Dialect/MQTOpt/Transforms/Passes.h"
+
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/Support/LLVM.h>
+#include <utility>
+
+namespace mqt::ir::opt {
+
+#define GEN_PASS_DEF_ELIDEPERMUTATIONS
+#include "mlir/Dialect/MQTOpt/Transforms/Passes.h.inc"
+
+/**
+ * @brief This pattern attempts to remove SWAP gates by re-ordering qubits.
+ */
+struct ElidePermutations final
+    : impl::ElidePermutationsBase<ElidePermutations> {
+
+  void runOnOperation() override {
+    // Get the current operation being operated on.
+    auto op = getOperation();
+    auto* ctx = &getContext();
+
+    // Define the set of patterns to use.
+    mlir::RewritePatternSet patterns(ctx);
+    populateElidePermutationsPatterns(patterns);
+
+    // Apply patterns in an iterative and greedy manner.
+    if (mlir::failed(APPLY_PATTERNS_GREEDILY(op, std::move(patterns)))) {
+      signalPassFailure();
+    }
+  }
+};
```

Comment:
Just out of curiosity: could you try and check whether this kind of boilerplate code is actually necessary? we keep writing that for all of our passes and it feels so repetitive that I can't help but think this could be automated by tablegen. Maybe tablegen has some convenient shortcuts for defining passes and/or patterns. 

Would be nice if you could read into that a little bit.

Reply:
I can have a look if I find anything, but I don't think it should become a part of this PR.

Reply:
I agree. This is supposedly a follow-up. The right pointers would probably be
- https://mlir.llvm.org/docs/DeclarativeRewrites/
- https://mlir.llvm.org/docs/PDLL/

Both look very promissing for reducing boilerplate code.